ideas/086-push-notif-v2/diagram.dot

```diff
@@ -1,33 +1,70 @@
 
-  node [shape=point fontsize=10];
-  edge [dir=none fontsize=10];
+  node [shape=point fontsize=10 fontname="Helvetica"];
+  edge [dir=none fontsize=10 fontname="Helvetica"];
 
   // Column labels
-  a [shape=Square label="Contact A"];
-  w [shape=Square label="Notification Server"];
-  b [shape=Square label="Contact B"];
+  a [shape=Square label="Contact A" style=filled color=lightblue];
+  w [shape=Square label="Arbitrary\nNotification Server"];
+  b [shape=Square label="Contact B" style=filled color=lightgreen];
 
   // Draw the 3 column headings, no line
   { rank=same; edge[style=invis] a -> w -> b   }
 
+  // Force lines to stay aligned
+  { rank=same; edge[style=invis] a2 -> w2 -> b2   }
+  { rank=same; edge[style=invis] a3 -> w3 -> b3   }
+  { rank=same; edge[style=invis] a4 -> w4 -> b4   }
+
   // Draw the columns
   a -> a1 [style=dotted weight=1000];
-  w -> w1 [style=dotted];
+  w -> w1 [style=dotted weight=1000];
   b -> b1 [style=dotted weight=1000];
 
-  a1 -> a2 -> a3 -> a4 -> a5 [weight=1000];
-  a3 [xlabel="Contact A wants\nto send a message\nto Contact B" weight=1000];
-  w1 -> w2 -> w3 -> w4 -> w5 [weight=1000];
-  b1 -> b2 -> b3 -> b4 -> b5 [weight=1000];
+  // Draw the vertical lines
+  a1 -> a2 -> a3 -> a4 [weight=1000] a4 -> a5 -> a6 [style=dotted weight=1000] a6 -> a7 [weight=1000] a7 -> a8 -> a9 [weight=1000] a9 -> amax [style=dotted weight=1000];
+  w2 -> w3 -> w4 [weight=1000] w4 -> w5 -> w6 [style=dotted weight=1000] w6 -> w7 -> w8 -> w9 [weight=1000] w9 -> wmax [style=dotted weight=1000];
+  b1 -> b2 -> b3 [style=dotted weight=1000] b3 -> b4 -> b5 -> b6 [style=dotted weight=1000] b6 -> b7 -> b8 -> b9 [weight=1000] b9 -> bmax [style=dotted weight=1000];
+  amax, wmax, bmax [style=invis];
 
   // Now each step in the ladder
-  { rank=same; a1 -> w1 [dir=forward label="Requests topic\nfor notifications\nto Contact B"]  }
-  { rank=same; a2 -> w2 [dir=back label="Replies with its Whisper topic"]  }
+  { rank=same; a1 -> w1 [dir=forward label="Discovery request\nover predefined topic (1)"]  }
+  w1 -> w2 [label="   Discovery and server negotiation\nrequests are encrypted\nasymetrically (2)" weight=1000];
+  b1, b2, b3, b4 [style=invis];
+
+  a2 -> w2 [dir=back label="Proposes itself"];
+
+  a3 -> w3 [dir=forward label="Accepts server"];
+
+  a4 -> w4 [dir=back label="Encrypted message\nusing subscription symkey\n(contains topic and new symkey)"]
+
+  a5 [label="" shape=none width=0 height=0 xlabel="At this point Contact A and\n Arbitrary Notification Server\ncan communicate securely,\nusing symkey provided by server"];
+  { rank=same; w5 -> b5 [dir=both label="Discovery request and negotiation,\nsame process as above for Contact A (...)"]  }
+
+  a6 [xlabel="Contact A starts chat\nwith Contact B"];
+  { rank=same; a6 -> w6 [dir=forward label="Requests key\nfor notifications\nto Contact B"]  }
+
+  { rank=same; a7 -> w7 [dir=back label="Replies with its key for Contact B"]  }
```

Comment:
I guess this key is generated when an account and a notification server agree on a topic, so, maybe the diagram could show when the notification server generates the key for each contact.

Reply:
Added:
![image](https://user-images.githubusercontent.com/138074/38855184-20235676-4223-11e8-8151-7e0080eba74e.png)